core/primitives/src/validator_mandates.rs

```diff
@@ -0,0 +1,259 @@
+    /// Initiates mandates corresponding to the provided `validators`. The validators must be sorted
+    /// by id in ascending order, so the validator with `ValidatorId` equal to `i` is given by
+    /// `validators[i]`.
+    ///
+    /// Only full mandates are assigned, partial mandates are dropped. For example, when the stake
+    /// required for a mandate is 5 and a validator has staked 12, then it will obtain 2 mandates.
+    pub fn new(config: ValidatorMandatesConfig, validators: &[ValidatorStake]) -> Self {
+        let num_mandates_per_validator: Vec<u16> =
+            validators.iter().map(|v| v.num_mandates(config.stake_per_mandate)).collect();
+        let num_total_mandates =
+            num_mandates_per_validator.iter().map(|&num| usize::from(num)).sum();
+        let mut mandates: Vec<ValidatorId> = Vec::with_capacity(num_total_mandates);
+
+        for i in 0..validators.len() {
+            for _ in 0..num_mandates_per_validator[i] {
+                // Each validator's position corresponds to its id.
+                mandates.push(i as ValidatorId);
+            }
+        }
+
+        let required_mandates = config.min_mandates_per_shard * config.num_shards;
+        if mandates.len() < required_mandates {
+            // TODO(chunk-validator-assignment) dynamically lower `stake_per_mandate` to reach enough mandates
+            panic!(
+                "not enough validator mandates: got {}, need {}",
+                mandates.len(),
+                required_mandates
+            );
+        }
+
+        Self { config, mandates }
+    }
+
+    /// Returns a validator assignment obtained by shuffling mandates.
+    ///
+    /// It clones mandates since [`ValidatorMandates`] is supposed to be valid for an epoch, while a
+    /// new assignment is calculated at every height.
+    ///
+    /// # Interpretation of the return value
+    ///
+    /// The returned vector contains one map per shard, with the position in the vector
+    /// corresponding to `shard_id` in `0..num_shards`.
+    ///
+    /// Each `HashMap` maps `ValidatorId`s to the number of mandates they have in the corresponding
+    /// shards. A validator whose id is not in a map has not been assigned to the shard.
+    ///
+    /// ## Example
+    ///
+    /// Let `res` be the return value of this function, then `res[0][1]` maps to the number of
+    /// mandates validator with `ValidatorId` 1 holds in shard with id 0.
+    pub fn sample<R>(&self, rng: &mut R) -> Vec<HashMap<ValidatorId, u16>>
+    where
+        R: Rng + ?Sized,
+    {
+        let shuffled_mandates = self.shuffle(rng);
+
+        // Assign shuffled seat at position `i` to the shard with id `i % num_shards`.
+        let mut assignments_per_shard = Vec::with_capacity(self.config.num_shards);
+        for shard_id in 0..self.config.num_shards {
+            let mut assignments = HashMap::new();
+            let mut idx = shard_id;
+
+            while idx < shuffled_mandates.len() {
```

Comment:
`for idx in (shard_id..shuffled_mandates.len()).step_by(self.config.num_shards)` ?

Reply:
Addressed in 1bd31df.